### PR TITLE
KT MERGING SOON: Workspace Stages 6.0

### PR DESCRIPTION
**Pull request by kevinmtrowbridge**
**@mmilano & QA folks: see the current state of this branch here: http://10.10.6.167:8080/**

This PR has to do with everything under the Workspace Stages heading of @mmilano's 6.0 spec document:
https://docs.google.com/document/d/1tii6TTcO_fsuMpeblvey9OURYnpDdZLV4z4FOt96Fqk/edit#heading=h.ncu35srt6xcd
Additionally here is the "negotiation" spreadsheet: https://docs.google.com/spreadsheets/d/1UDY64Ni-v14UvIwin70KujUTfMFYLSjYlQKWjEvuYI0/edit#gid=0

These Jira stories are related: DEV-14258 DEV-14277

### CODE REVIEW

**Comment by kevinmtrowbridge** [comment]
Still in progress, but the bones of the work are present, with tests.

#### Core modifications:
* add columns to Workspaces table: 'stage' & 'stage_reason'
* add new event "WorkspaceStageChanged"
* current workspaces are set to "modeling", new workspaces will have nil stage currently
* update tests
    
#### Api modifications:
* modify workspaces presenter to pass down these new attributes
* update tests
    
#### Frontend modifications:
* pipe value into backbone views
* appears everywhere except the typeahead search
* ability to edit workspace stage
* update tests

### TODO

* No styling has been done, just the bare "stage" value is present.  Share work with @mmilano & work with him on styling.
* Styling the search results is a tricky area.  I am still working on this, have not pushed these changes yet.
* I think some of the new Jasmine tests may fail.  Curious to see what Travis has to say.

**Comment by kevinmtrowbridge** [comment]
@mmilano & QA folks: I can see some obvious bugs there already (the WorkspaceStageChanged event isn't showing up perhaps because of caching?), but in the interests of greater communication, you can see the current state of this branch here: http://10.10.6.167:8080/

This is a snapshot of the dev branch deployed using docker.

Here are a few screenshots of the current state:

alpine chorus

alpine chorus-1

alpine chorus-2

**Comment by mmilano** [comment]
@kevinmtrowbridge based on initial glance, i have amended the design doc to account for a design gap. namely, new workspaces.
for now, new workspaces should be given the first stage by default. Not "nil."

the design spec document now includes this bit:
> New Workspaces
> New workspaces are given the first stage by default.

this means that all workspaces have to have a value for this field.

for longer view - not H priority for R1 -  there is also a backlog story about allowing the user to choose the value in the 'new workspace' dialog.

**Comment by mmilano** [comment]
@kevinmtrowbridge the workspace stage names and description text needs to use the values that we will add to the common language file (messages_en.properties).

**Comment by mmilano** [comment]
I have added to the branch with a set of shared styles
* there is now a shared 'workspace stage badge' component that can be used across the app
* the workspace stage badge has some preliminary styles

File: components/core/app/models/workspace.rb [hunk 1]
```diff
@@ -8,9 +8,14 @@ class Workspace < ActiveRecord::Base
 
   PROJECT_STATUSES = [:on_track, :needs_attention, :at_risk]
 
+  STAGES = {'exploration' => {:name => 'Exploration', :description => 'Exploring data science.'},
+            'modeling' => {:name => 'Modeling', :description => 'Taking the next steps.'},
+            'refinement' => {:name => 'Refinement', :description => 'Making it real.'},
+            'production' => {:name => 'Production', :description => 'Now it is useful!'}}
```
Comment (jleealpine) [inline components/core/app/models/workspace.rb]: I assume these descriptions are placeholders til the real deal?

Comment (kevinmtrowbridge) [inline components/core/app/models/workspace.rb]: >  descriptions are placeholders
Yes I just made those up for now.

> workspace stage names and description text needs to use the values that we will add to the common language file (messages_en.properties)
I will move the text into the Frontend's `messages_en.properties` file and just store the key in the Core database.

Comment (mmilano) [inline components/core/app/models/workspace.rb]: i added keys for these in the addendum from a few hours ago today.

**Comment by jleealpine** [comment]
Out of curiosity, are there any plans either immediate or in the future to remove the "project status" feature on workspaces?  I ask just because the implementation for the two are very similar, but understand that it might be worth having both because they serve somewhat different purposes.

**Comment by kevinmtrowbridge** [comment]
> remove the "project status" feature on workspaces

It's important to delete things!  We should seriously discuss deleting that.  For now I have just tried to keep this branch atomic, and figured we could make that decision and remove that feature in another PR.  But I defer the decision to @mmilano.

**Comment by kevinmtrowbridge** [comment]
> new workspaces should be given the first stage by default. Not "nil."

I have this taken care of now.

> workspace stage names and description text needs to use the values that we will add to the common language file (messages_en.properties)

It now stores only the "stage" value in the database and pulls the names and description text from the messages_en.properties file.

> there is now a shared 'workspace stage badge' component that can be used across the app

Thank you @mmilano!  I am now using the workspaceStageBadge component everywhere.

Additionally, on the WorkspaceShow page the the WorkspaceActivityList now updates when you update the stage or the status.

-----

I haven't pushed my changes yet.  I will now move on to:

* styling the search results
* getting the tests passing

**Comment by kevinmtrowbridge** [comment]
I think this branch is ready for some more fine-grained and critical review.

**@mmilano & QA folks: see the current state of this branch here: http://10.10.6.167:8080/**

Here are some screenshots of the current state of this branch:

Workspace card:
alpine chorus-2

Workspace show page or homepage:
alpine chorus-4

Search results:
alpine chorus-5

Workspaces index page:
alpine chorus-6

File: components/frontend/app/assets/translations/messages_en.properties [hunk 2]
```diff
@@ -2386,6 +2387,7 @@ workspace.delete.toast=The workspace <span class="message_callout">{{workspaceNa
 
 workspace.edit_members=Add or Edit Members
 actions.edit_workspace_members=Add or Edit Members
+actions.edit_workspace_stage=Edit Stage
```
Comment (kevinmtrowbridge) [inline components/frontend/app/assets/translations/messages_en.properties]: Whoops, I believe this should be 'Update Stage' ...

File: components/frontend/app/assets/translations/messages_en.properties [hunk 3]
```diff
@@ -2442,10 +2444,25 @@ workspace.project.status.at_risk=At Risk
...
 
+workspace.stage.exploration.label=Exploration
+workspace.stage.modeling.label=Modeling
+workspace.stage.refinement.label=Refinement
+workspace.stage.production.label=Production
+
+workspace.stage.exploration.description=exploration workspace stage description blah blah blah placeholder
+workspace.stage.modeling.description=modeling workspace stage description blah blah blah placeholder
+workspace.stage.refinement.description=refinement workspace stage description blah blah blah longer placeholder placeholder longer placeholder
+workspace.stage.production.description=production workspace stage description blah blah blah placeholder
```
Comment (kevinmtrowbridge) [inline components/frontend/app/assets/translations/messages_en.properties]: @mmilano: please update these values with more correct text when you have it.

Comment (mmilano) [inline components/frontend/app/assets/translations/messages_en.properties]: noted.
dont worry about the text right now, however. the important thing at this stage is that the right keys are there and that they are used. 
we will be revising the text up until release day.

**Comment by kevinmtrowbridge** [comment]
@mmilano: I will be merging this branch in, as soon as the tests pass.  Let me know if I should not do this.  Plan is to let you and Josh edit the "blah blah" workspace stage descriptions, in master.

**Comment by mmilano** [comment]
@kevinmtrowbridge sounds fine. go for it.

**Pull request merged**